pkg/webhooks/server.go

```diff
@@ -360,7 +374,61 @@ func (ws *WebhookServer) resourceMutation(request *v1beta1.AdmissionRequest) *v1
 
 func (ws *WebhookServer) resourceValidation(request *v1beta1.AdmissionRequest) *v1beta1.AdmissionResponse {
 	logger := ws.log.WithName("resourceValidation").WithValues("uid", request.UID, "kind", request.Kind.Kind, "namespace", request.Namespace, "name", request.Name, "operation", request.Operation)
+	checked, err := userinfo.IsRoleAuthorize(ws.rbLister, ws.crbLister, ws.rLister, ws.crLister, request)
+	if err != nil {
+		logger.Error(err, "failed to get RBAC infromation for request")
+	}
+
+	if !checked {
+		if request.Operation == v1beta1.Delete || request.Operation == v1beta1.Update {
+			// convert RAW to unstructured
+			var resource *unstructured.Unstructured
+			if request.Operation == v1beta1.Delete {
+				resource, err = enginutils.ConvertToUnstructured(request.OldObject.Raw)
+			} else {
+				resource, err = enginutils.ConvertToUnstructured(request.Object.Raw)
+			}
+			if err != nil {
+				//TODO: skip applying the admission control ?
+				logger.Error(err, "failed to convert RAR resource to unstructured format")
 
+				return &v1beta1.AdmissionResponse{
+					Allowed: false,
+					Result: &metav1.Status{
+						Status:  "Failure",
+						Message: err.Error(),
+					},
+				}
+			}
+
+			oldResource, err := ws.client.GetResource(resource.GetKind(), resource.GetNamespace(), resource.GetName())
```

Comment:
Why do you want to get the resource using the client here? Won't unstructured `resource` represent the resource object already?

Reply:
it is possible that the request doesn't not have label. we can only trust etcd data

Reply:
Well.. this is on resource UPDATE / DELETE, the raw object of the request should be consistent with etcd data right?